File: src/main/java/dev/hoot/api/game/Game.java
package dev.hoot.api.game;

import dev.hoot.api.movement.pathfinder.GlobalCollisionMap;
import dev.hoot.api.movement.pathfinder.RegionManager;
import dev.hoot.api.movement.pathfinder.Walker;
import dev.hoot.api.widgets.Tab;
import dev.hoot.api.widgets.Tabs;
import dev.hoot.api.widgets.Widgets;
import meteor.Logger;
import meteor.Main;
import net.runelite.api.Client;
import net.runelite.api.GameState;
import net.runelite.api.packets.ClientPacket;
import net.runelite.api.widgets.Widget;
import net.runelite.api.widgets.WidgetID;
import net.runelite.api.widgets.WidgetInfo;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.zip.GZIPInputStream;

public class Game
{
	private static final int MEMBER_DAYS_VARP = 1780;
	private static final int CUTSCENE_VARBIT = 542;
	private static final String LOGOUT_ACTION = "Logout";
	private static ClientPacket clientPacket = null;
	private static GlobalCollisionMap globalCollisionMap = null;

	public static Client getClient()
	{
		return Main.client;
	}

	public static ClientPacket getClientPacket()
	{
		if (clientPacket == null)
			clientPacket = Main.INSTANCE.getClient().createClientPacket(-1, -1);
		return clientPacket;
	}

	public static GlobalCollisionMap getGlobalCollisionMap()
	{
		if (globalCollisionMap == null) {
			try {
				return new GlobalCollisionMap(
						new GZIPInputStream(
								new ByteArrayInputStream(
										ClassLoader.getSystemClassLoader().getResourceAsStream("regions").readAllBytes()
								)
						).readAllBytes()
				);
			} catch (IOException ex) {
				ex.printStackTrace();
			}
		}
		return globalCollisionMap;
	}

	public static boolean isLoggedIn()
	{
		return getState() == GameState.LOGGED_IN || getState() == GameState.LOADING;
	}

	public static boolean isOnLoginScreen()
	{
		return getState() == GameState.LOGIN_SCREEN
				|| getState() == GameState.LOGIN_SCREEN_AUTHENTICATOR;
	}

	public static GameState getState()
	{
		return getClient().getGameState();
	}

	public static int getWildyLevel()
	{
		Widget wildyLevelWidget = Widgets.get(WidgetInfo.PVP_WILDERNESS_LEVEL);
		if (!Widgets.isVisible(wildyLevelWidget))
		{
			return 0;
		}

		// Dmm
		if (wildyLevelWidget.getText().contains("Guarded")
				|| wildyLevelWidget.getText().contains("Protection"))
		{
			return 0;
		}

		if (wildyLevelWidget.getText().contains("Deadman"))
		{
			return Integer.MAX_VALUE;
		}
		String widgetText = wildyLevelWidget.getText();
		if (widgetText.indexOf('<') == -1)
			return 0;
		return Integer.parseInt(widgetText.substring(0, widgetText.indexOf('<')).replace("Level: ", ""));
	}

	public static int getMembershipDays()
	{
		return Vars.getVarp(MEMBER_DAYS_VARP);
	}

	public static boolean isInCutscene()
	{
		return Vars.getBit(CUTSCENE_VARBIT) > 0;
	}

	public static void logout()
	{
		Widget logOutHopper = Widgets.get(WidgetID.WORLD_SWITCHER_GROUP_ID, x -> x.hasAction(LOGOUT_ACTION));
		if (logOutHopper != null)
		{
			logOutHopper.interact(LOGOUT_ACTION);
			return;
		}

		Widget logOut = Widgets.get(WidgetID.LOGOUT_PANEL_ID, x -> x.hasAction(LOGOUT_ACTION));
		if (logOut != null)
		{
			logOut.interact(LOGOUT_ACTION);
			return;
		}

		if (!Tabs.isOpen(Tab.LOG_OUT))
		{
			Tabs.open(Tab.LOG_OUT);
		}
	}
}

File: src/main/kotlin/meteor/api/game/Game.kt
package meteor.api.game

import Main.client
import meteor.api.movement.pathfinder.GlobalCollisionMap
import meteor.api.movement.pathfinder.RegionManager
import meteor.api.widgets.Tab
import meteor.api.widgets.Tabs
import meteor.api.widgets.Widgets
import net.runelite.api.GameState
import net.runelite.api.packets.ClientPacket
import net.runelite.api.widgets.Widget
import net.runelite.api.widgets.WidgetID
import net.runelite.api.widgets.WidgetInfo
import java.io.ByteArrayInputStream
import java.net.URL
import java.util.zip.GZIPInputStream


object Game {
    private const val MEMBER_DAYS_VARP = 1780
    private const val CUTSCENE_VARBIT = 542
    private const val LOGOUT_ACTION = "Logout"

    val clientPacket: ClientPacket? = null
    val regions = URL(RegionManager.API_URL + "/regions").openStream()
    val globalCollisionMap = GlobalCollisionMap(GZIPInputStream(ByteArrayInputStream(regions.readAllBytes())).readAllBytes())

    val isLoggedIn: Boolean
        get() = state == GameState.LOGGED_IN || state == GameState.LOADING
    val isOnLoginScreen: Boolean
        get() = (state == GameState.LOGIN_SCREEN
                || state == GameState.LOGIN_SCREEN_AUTHENTICATOR)
    val state: GameState
        get() = client!!.gameState

    // Dmm
    val wildyLevel: Int
        get() {
            val wildyLevelWidget: Widget? = Widgets.get(WidgetInfo.PVP_WILDERNESS_LEVEL)
            if (wildyLevelWidget == null || GameThread.invokeLater { wildyLevelWidget.isHidden }!!) {
                return 0
            }

            // Dmm
            if (wildyLevelWidget.text.contains("Guarded")
                || wildyLevelWidget.text.contains("Protection")
            ) {
                return 0
            }
            return if (wildyLevelWidget.text.contains("Deadman")) {
                Int.MAX_VALUE
            } else wildyLevelWidget.text.replace("Level: ", "").toInt()
        }
    val membershipDays: Int
        get() = Vars.getVarp(MEMBER_DAYS_VARP)
    val isInCutscene: Boolean
        get() = Vars.getBit(CUTSCENE_VARBIT)!! > 0

    fun logout() {
        val logOutHopper: Widget? = Widgets.get(WidgetID.WORLD_SWITCHER_GROUP_ID) { x -> x!!.hasAction(LOGOUT_ACTION) }
        if (logOutHopper != null) {
            logOutHopper.interact(LOGOUT_ACTION)
            return
        }
        val logOut: Widget? = Widgets.get(WidgetID.LOGOUT_PANEL_ID) { x -> x!!.hasAction(LOGOUT_ACTION) }
        if (logOut != null) {
            logOut.interact(LOGOUT_ACTION)
            return
        }
        if (!Tabs.isOpen(Tab.LOG_OUT)) {
            Tabs.open(Tab.LOG_OUT)
        }
    }
}